CSS tree counting functions - `sibling-index()` and `sibling-count()` (#40120)
mdn/content: files/en-us/web/css/sibling-index/index.md

---
title: sibling-index()
slug: Web/CSS/sibling-index
page-type: css-function
browser-compat: css.types.sibling-index
---

{{CSSRef}}

> The **`sibling-index()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Values_and_Units/CSS_Value_Functions) represents an integer of the current element in the DOM tree on which the notation is used among its siblings. The function returns the contextual child index as a number. In other words, the `sibling-index()` function returns a number of the child element according to its position among all the sibling elements within a parent element.

Reviewer: ```suggestion
The **`sibling-index()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Values_and_Units/CSS_Value_Functions) returns an integer representing the position of the current element in the DOM tree relative to all its sibling elements. The returned value is the index number of the contextual child's  position among all the sibling elements within a parent element, with the first child returning `1` and the last child, returning {{domxref("Element.children")}}.length.
```
or something like that.

> [!NOTE]
> Like the {{CSSxRef(":nth-child()")}} pseudo-class, `sibling-index()` starts from 1, not 0.

## Syntax

```css
li {
  width: calc(sibling-index() * 100px);
}
```

### Parameters

The `sibling-index()` function doesn't accept parameters.

### Return value

> A number representing the current element in the DOM tree among its siblings.

Reviewer: ```suggestion
An integer, the position of the current element in the DOM tree's sibling order.
```

## Examples

### Dynamic list width

For example, set the list item width dynamically based on the `<li>` elements order in a list.

#### HTML

```html
<ul>
  <li>One</li>
  <li>Two</li>
  <li>Three</li>
  <li>Four</li>
</ul>
```

#### CSS

```css
li {
  width: calc(sibling-index() * 50px);
  background-color: #faa;
}
```

### Ordered List

> Now let's create an alternative ordered list using `sibling-index()`.

Reviewer: why would we do this? There's an ordered list `<ol>`. Let's come up with something that is not bad practice. I think we could actually just skip this one as the last one is good. Just need to add "results" and a macro to that and the previous example

---
title: sibling-index()
slug: Web/CSS/sibling-index
page-type: css-function
browser-compat: css.types.sibling-index
---

{{CSSRef}}

The **`sibling-index()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Values_and_Units/CSS_Value_Functions) returns an integer representing the position of the current element in the DOM tree relative to all its sibling elements. The returned value is the index number of the contextual child's position among all the sibling elements within a parent element, with the first child returning `1` and the last child, returning {{domxref("Element.children")}}.length.

> [!NOTE]
> Like the {{CSSxRef(":nth-child()")}} pseudo-class, `sibling-index()` starts from 1, not 0.

## Syntax

```css
li {
  width: calc(sibling-index() * 100px);
}
```

### Parameters

The `sibling-index()` function doesn't accept parameters.

### Return value

> An integer, the position of the current element in the DOM tree's sibling order.

Reviewer: ```suggestion
An integer; the position of the current element in the DOM tree's sibling order.
```

## Examples

### Dynamic list width

This example demonstrates how to dynamically increase the width of each {{htmlelement("li")}} item in the {{htmlelement("ul")}} by `50px`.

#### HTML

```html
<ul>
  <li>One</li>
  <li>Two</li>
  <li>Three</li>
  <li>Four</li>
</ul>
```

#### CSS

```css
li {
  width: calc(sibling-index() * 50px);
  background-color: #faa;
}
```

#### Results

{{EmbedLiveSample("Dynamic list width", "300", "100")}}

### Ordered List

> This example demonstrates how to create an ordered list using `sibling-index()`, without using the {{htmlelement("ol")}} element.
> 
> Because not always you have the ability to change the HTML. In these cases, you so you can do that with CSS.

Reviewer: ```suggestion
This example demonstrates how to create an ordered list using `sibling-index()`, without using the {{htmlelement("ol")}} element. Always use the most semantic element for the context; this example is included to  demonstrate what can be done with CSS when you have the ability to change the HTML.
```
encouraging people to not do this

#### HTML

When you have {{htmlelement("nav")}} container and several children {{htmlelement("div")}} elements.

```html
<nav arial-label="Ordered list">
  <div>One</div>
  <div>Two</div>
  <div>Three</div>
  <div>Four</div>
</nav>
```

#### CSS

> You can transform it to a numbered list by disapling the sibling-index before each {{htmlelement("div")}} element using the {{CSSxRef("before")}} pseudo-element with the `sibling-index()` value.

Reviewer: ```suggestion
We make it visually appear as a numbered list by displaying the sibling-index before each {{htmlelement("div")}} element using the {{CSSxRef("before")}} pseudo-element, setting the {{cssxref("content")}} to be the integer returned by the `sibling-index()` function.
```

```css
div {
  --list-index: sibling-index();
  display: flex;
  gap: 1ch;
}

div::before {
  content: var(--list-index);
}
```

#### Results

{{EmbedLiveSample("Ordered List", "300", "100")}}

### Sequential animations

Combining `sibling-index()` with CSS animations opens new possibilities. In this example, the opacity of elements in sequential order by setting an {{cssxref("animation-delay")}} based on their order in the DOM.

#### HTML

We include a container element with four children:

```html
<ul>
  <li>One</li>
  <li>Two</li>
  <li>Three</li>
  <li>Four</li>
</ul>
```

#### CSS

We apply the `fade-in` animation to each element. We use the `sibling-index()` function within a {{cssxref("calc()")}} function to set the duration of the `animation-delay` based on the source element's position in the source order. The {{cssxref("animation-fill-mode")}} applies the animation's `0%` keyframe until the `animation-duration` expires.

```css
ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

li {

>   animation-name: fade;
>   animation-duration: 1s;
>   animation-iteration-count: 1;
>   animation-timing-function: linear;
>   animation-fill-mode: backwards;
>   animation-delay: calc(1s * sibling-index());
> }
> 
> @keyframes fade {

Reviewer: ```suggestion
  animation-name: fade-in;
  animation-duration: 1s;
  animation-iteration-count: 1;
  animation-timing-function: linear;
  animation-fill-mode: backwards;
  animation-delay: calc(1s * sibling-index());
}

@keyframes fade-in {
```
